components/server/src/user/user-service.spec.db.ts

```diff
@@ -4,51 +4,35 @@
  * See License.AGPL.txt in the project root for license information.
  */
 
-import { DBUser, TypeORM, UserDB } from "@gitpod/gitpod-db/lib";
+import { Experiments } from "@gitpod/gitpod-protocol/lib/experiments/configcat-server";
+import * as chai from "chai";
 import { Container } from "inversify";
-import { suite, test } from "@testdeck/mocha";
-import { DBTeam } from "@gitpod/gitpod-db/lib/typeorm/entity/db-team";
+import { createTestContainer } from "../test/service-testing-container-module";
 import { UserService } from "./user-service";
-import * as chai from "chai";
-import { dbContainerModule } from "@gitpod/gitpod-db/lib/container-module";
-import { productionContainerModule } from "../container-module";
-import { Config } from "../config";
+import { DBUser, TypeORM } from "@gitpod/gitpod-db/lib";
 
 const expect = chai.expect;
 
-const testContainer = new Container();
-testContainer.load(dbContainerModule());
-testContainer.load(productionContainerModule);
-testContainer.rebind(Config).toConstantValue({
-    blockNewUsers: {
-        enabled: false,
-    },
-} as any);
-
-@suite
-class UserServiceSpec {
-    userDB = testContainer.get<UserDB>(UserDB);
-
-    async before() {
-        await this.wipeRepos();
-    }
+describe("UserService", async () => {
+    let container: Container;
+    let userService: UserService;
 
-    async after() {
-        await this.wipeRepos();
-    }
-
-    async wipeRepos() {
-        const typeorm = testContainer.get<TypeORM>(TypeORM);
-        const mnr = await typeorm.getConnection();
-        await mnr.getRepository(DBUser).delete({});
-        await mnr.getRepository(DBTeam).delete({});
-    }
+    beforeEach(async () => {
+        container = createTestContainer();
+        Experiments.configureTestingClient({
+            centralizedPermissions: true,
+        });
```

Comment:
How does this ensure it doesn't interfere with other tests? Is it due to NodeJS running single-threaded that we're sure we're not affecting other tests?

Reply:
The container and all its state is per test. So that is not a problem. The database however is not, so we have issues here. I was looking into applying Resource Acquisition Is Initialization" (RAII) principle and having a test cleaning up only their resources which would reduce the possibility of interference of parallel running tests. But we have some methods that would list resources globally which are hard to test when some other service could add or remove their resources in the meantime. I believe we should keep running these tests in parallel, but maybe give the whole suite its own instance, so at least we could run the usage tests in parallel (which they currently do)